### PR TITLE
[AMDGPU] Do not run GCNNSAReassign pass for GFX12

### PR DESCRIPTION
**Pull request by jayfoad**
GFX12 does not have separate NSA and non-NSA encodings.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/GCNNSAReassign.cpp [hunk 1]
```diff
@@ -237,7 +237,7 @@ GCNNSAReassign::CheckNSA(const MachineInstr &MI, bool Fast) const {
 
 bool GCNNSAReassign::runOnMachineFunction(MachineFunction &MF) {
   ST = &MF.getSubtarget<GCNSubtarget>();
-  if (!ST->hasNSAEncoding())
+  if (!ST->hasNSAEncoding() || ST->getGeneration() > GCNSubtarget::GFX11)
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNNSAReassign.cpp:240]: If it doesn't have NSA encodings, why isn't it covered by !ST->hasNSAEncoding()?

Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/GCNNSAReassign.cpp:240]: It _only_ has NSA encoding, so there is no point in running a pass that tries to optimize it back to a non-NSA encoded form (I guess we could add a `hasNonNSAEncoding` hook for that).

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/GCNNSAReassign.cpp:240]: Better to hide it in the named subtarget predicate

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/GCNNSAReassign.cpp (hunk 1 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/GCNNSAReassign.cpp]: Done.

**Pull request merged**